### PR TITLE
add nft-voter-v2 (support cNFT)

### PR DESCRIPTION
**Pull request by hauchu1998**
integrate nft-voter-v2 program (only for devnet)
nft-voter-v2 address: `GnftVc21v2BRchsRa9dGdrVmJPLZiRHe9j2offnFTZFg`
nft-voter-v2 support compressed NFTs to be voting powers in a realm.
Please reference [here](https://github.com/solana-labs/interns-codehub/tree/master/cNft-Governance/governance-program-library/programs/nft-voter) for more information.

governance-ui changes:
1. add a flag `ON_NFT_VOTER_V2` to switch between v1(current) and v2. (manually change for now, might add a switch button afterwards?)
2. add `getUpdateVoterWeightInstruction` and `getUpdateVoterWeightInstructionV2` to distinguish the difference, as well as `getCastNftVoteInstruction` and `getCastNftVoteInstructionV2`
3. change the instruction flow for each voter action.

### CODE REVIEW

File: .prettierrc [hunk 1]
```diff
@@ -1,4 +1,5 @@
 {
   "semi": false,
-  "singleQuote": true
-}
+  "singleQuote": true,
+  "trailingComma": "es5"
```
Comment (asktree) [inline .prettierrc:4]: ?

File: NftVotePlugin/accounts.ts [hunk 2]
```diff
@@ -9,6 +13,14 @@ export interface NftVoteRecord {
   publicKey: PublicKey
 }
 
+export interface NftActionTicket {
```
Comment (asktree) [inline NftVotePlugin/accounts.ts:16]: does anchor + IDL not provide this for us?

File: NftVotePlugin/accounts.ts [hunk 3]
```diff
@@ -45,3 +57,53 @@ export const getNftVoteRecordProgramAddress = async (
...
+          offset: 40,
+          bytes: voter.toBase58(),
+        },
+      },
+    ]
+  )) as unknown) as NftActionTicket[]
+  return nftActionTicketsFiltered
+}
+
+export const getNftActionTicketProgramAddress = async (
```
Comment (asktree) [inline NftVotePlugin/accounts.ts:85]: i thought anchor had stuff for this? 
note: i don't actually use the official anchor sdk, I always used anchor-client-gen, so i really dont know.

not a big deal either way

File: .prettierrc (hunk 1 above)
Comment (hauchu1998) [reply .prettierrc:4]: Oh my bad, I'll remove this.

File: utils/uiTypes/VotePlugin.ts [hunk 4]
```diff
@@ -248,25 +261,35 @@ export class VotingClient {
...
+          updateVoterWeightRecordIx,
+        } = await getUpdateVoterWeightRecordInstructionV2(
+          this.client,
+          walletPk,
+          registrar,
+          voterWeightPk,
+          this.votingNfts,
+          type
+        )
+        createNftActionTicketIxs?.push(...createNftTicketIxs)
```
Comment (asktree) [inline utils/uiTypes/VotePlugin.ts:289]: why is `createNftActionTicketIxs` needed instead of just adding it to `instructions`?

Comment (asktree) [reply utils/uiTypes/VotePlugin.ts:289]: i guess it just makes it easier to batch the txs how you want? whatever, it doesnt really matter, this file is way past the point of no return on code quality anyway

File: actions/chat/postMessage.ts [hunk 5]
```diff
@@ -54,8 +62,42 @@ export async function postChatMessage(
...
+
+  const instructionsChunks = [
+    ...nftTicketAccountsChuncks.map((txBatch, batchIdx) => {
+      return {
+        instructionsSet: txBatchesToInstructionSetWithSigners(
+          txBatch,
+          [],
+          batchIdx
+        ),
+        sequenceType: SequenceType.Parallel,
```
Comment (asktree) [inline actions/chat/postMessage.ts:78]: hmm i think the way you are doing this will make it such that the post message stuff might happen too soon relative to the ticket creation, due to sequencetype parallel here

File: components/Members/MembersTabs.tsx [hunk 6]
```diff
@@ -17,12 +17,14 @@ interface MembersTabsProps {
...
   onChange: (x) => void
   tabs: Array<Member>
+  isNftMode?: boolean
 }
 
 const MembersTabs: FunctionComponent<MembersTabsProps> = ({
   activeTab,
   onChange,
   tabs,
+  isNftMode,
```
Comment (asktree) [inline components/Members/MembersTabs.tsx:27]: this doesn't really need to be a prop because this component can just use hooks and figure this out on its own, but whatever

File: components/Members/MembersTabs.tsx [hunk 7]
```diff
@@ -31,6 +33,8 @@ const MembersTabs: FunctionComponent<MembersTabsProps> = ({
     ? tokenPriceService.getTokenInfo(realm?.account.communityMint.toBase58())
         ?.symbol
     : ''
+
+  const nftName = isNftMode ? 'NFT' : ''
```
Comment (asktree) [inline components/Members/MembersTabs.tsx:37]:
```suggestion
  const nftName = isNftMode ? 'NFT' : undefined
```

File: components/Members/useMembers.tsx [hunk 8]
```diff
@@ -212,7 +256,7 @@ export const useMembersQuery = () => {
       console.log('useMembers is fetching')
 
       let members = [...membersWithTokensDeposited]
-
+      console.log(members.map((x) => x.communityVotes.toString()))
```
Comment (asktree) [inline components/Members/useMembers.tsx:259]: remove console log

File: components/Members/useMembers.tsx [hunk 9]
```diff
@@ -224,7 +268,7 @@ export const useMembersQuery = () => {
       const activeMembers = members.filter(
         (x) => !x.councilVotes.isZero() || !x.communityVotes.isZero()
       )
-
+      console.log(activeMembers.map((x) => x.communityVotes.toString()))
```
Comment (asktree) [inline components/Members/useMembers.tsx:271]: remove console log (or at least label it)

File: components/ProposalTopVotersNftChart.tsx [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+interface Props {
+  showNfts: boolean
+  isNftMode: boolean
+  className?: string
+  voteType?: number
+  highlighted?: string
+}
+
+const getContents = (css, isLoading, highlighted, nfts, voteType) => {
```
Comment (asktree) [inline components/ProposalTopVotersNftChart.tsx:15]: this function should not exist, nor should any similar function ever exist, the jsx should just be in the component

File: hooks/useNftRegistrarCollection.ts [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    s.state.nftMintRegistrar,
+  ])
+
+  return useMemo(
+    () =>
+      (currentPluginPk &&
+        NFT_PLUGINS_PKS.includes(currentPluginPk?.toBase58()) &&
+        nftMintRegistrar?.collectionConfigs.map((x) =>
+          x.collection.toBase58()
+        )) ||
```
Comment (asktree) [inline hooks/useNftRegistrarCollection.ts:19]:
```suggestion
        )) ??
```

File: hooks/useVoteRecords.ts [hunk 12]
```diff
@@ -163,5 +197,71 @@ export default function useVoteRecords(proposal?: ProgramAccount<Proposal>) {
   ])
   ///////
 
+  // for nft-voter
+  useEffect(() => {
```
Comment (asktree) [inline hooks/useVoteRecords.ts:201]: use useAsync or useQuery, not useEffect

Comment (asktree) [reply hooks/useVoteRecords.ts:201]: also i know this code is in the hell zone and you are just copying patterns, but more comments would be nice just explaining whats going on (what's this data? why are we getting it?) because its not the easiest to follow if you are low brainpower like me

File: hub/components/NewWallet/useNewWalletTransaction.ts [hunk 13]
```diff
@@ -79,8 +86,25 @@ const useNewWalletCallback = (
       wallet.publicKey,
     );
 
+    // the list will have elements only if the plugin is NFTVoterClient2
+    // so we can just add the instructions to transactionInstructions
+    const createGovernanceAccountsChunks = chunks(
```
Comment (asktree) [inline hub/components/NewWallet/useNewWalletTransaction.ts:91]: this seems mislabeled

File: hub/components/NewWallet/useNewWalletTransaction.ts [hunk 14]
```diff
@@ -51,12 +56,14 @@ const useNewWalletCallback = (
     );
 
     const instructions: TransactionInstruction[] = [];
+    const createCreateGovernanceTicketIxs: TransactionInstruction[] = [];
```
Comment (asktree) [inline hub/components/NewWallet/useNewWalletTransaction.ts:59]: it could just be called createTicketIxs or whatever. its clear from context we are making a new governance. also adding all this extra code, just for nft-voter, any time we do anything is kinda lame. i think its because you are trying to make ticket creation parallel for efficiency reasons, but im not sure thats even how the thing works because i dont think the sequential ones wait for the parallel ones.

File: hub/providers/Proposal/createProposal.ts [hunk 15]
```diff
@@ -30,9 +31,15 @@ import {
 } from 'actions/createProposal';
 import { tryGetNftRegistrar } from 'VoteStakeRegistry/sdk/api';
 
+import {
```
Comment (asktree) [inline hub/providers/Proposal/createProposal.ts:34]: `hubs` folder is not maintained so no changes in this file should matter; sorry about that

File: hub/components/NewWallet/useNewWalletTransaction.ts (hunk 13 above)
Comment (asktree) [reply hub/components/NewWallet/useNewWalletTransaction.ts:91]: that said hubs folder is not maintained anyway

File: hub/components/NewWallet/useNewWalletTransaction.ts (hunk 14 above)
Comment (asktree) [reply hub/components/NewWallet/useNewWalletTransaction.ts:59]: however hubs folder is not actually maintained

File: models/proposal/buildTopVoters.ts [hunk 16]
```diff
@@ -2,16 +2,16 @@ import {
...
+  // Proposal,
   VoteKind,
 } from '@solana/spl-governance'
 import { MintInfo } from '@solana/spl-token'
 import BN from 'bn.js'
 import { PublicKey } from '@solana/web3.js'
 import { BigNumber } from 'bignumber.js'
 
-import { calculateMaxVoteScore } from '@models/proposal/calulateMaxVoteScore'
+// import { calculateMaxVoteScore } from '@models/proposal/calulateMaxVoteScore'
```
Comment (asktree) [inline models/proposal/buildTopVoters.ts:14]: just delete imports that arent used anymore. they are easy to bring back!

File: pages/dao/[symbol]/members/Members.tsx [hunk 17]
```diff
@@ -16,14 +16,15 @@ import PaginationComponent from '@components/Pagination'
 import useWalletOnePointOh from '@hooks/useWalletOnePointOh'
 import { useMembersQuery } from '@components/Members/useMembers'
 
-const Members = () => {
+const Members = ({ isNftMode }: { isNftMode: boolean }) => {
```
Comment (asktree) [inline pages/dao/[symbol]/members/Members.tsx:19]: lets not prop drill `isNftMode`. components can just figure it out for themselves and act accordingly.

File: tools/cnftParams.ts [hunk 18]
```diff
@@ -0,0 +1,100 @@
+// import { PublicKey } from "@solana/web3.js"
+import { Connection, PublicKey } from '@solana/web3.js'
+import { ConcurrentMerkleTreeAccount } from '@solana/spl-account-compression'
+import * as bs58 from 'bs58'
+import { fetchDasAssetProofById } from '@hooks/queries/digitalAssets'
+import { getNetworkFromEndpoint } from '@utils/connection'
+import * as anchor from '@coral-xyz/anchor'
+
+class AccountData {
```
Comment (asktree) [inline tools/cnftParams.ts:9]: whats up with this file? why would we make this class?

File: tools/cnftParams.ts [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+
+function bufferToArray(buffer: Buffer): number[] {
+  const nums: number[] = []
+  for (let i = 0; i < buffer.length; i++) {
+    nums.push(buffer[i])
+  }
+  return nums
+}
+
+export async function getCompressedNftParamAndProof(
```
Comment (asktree) [inline tools/cnftParams.ts:36]: my understanding is that this is supposed to be creating the compressed nft param for use with nft voter v2 program. a docstring comment should explain this, otherwise this file is pure mystery for anyone who isn't reviewing a PR about nft voter v2

File: pages/dao/[symbol]/proposal/[pk]/explore.tsx [hunk 20]
```diff
@@ -26,9 +30,30 @@ export default function Explore() {
...
   const endpoint = connection.endpoint
 
-  const isMulti = proposal?.account.voteType !== VoteType.SINGLE_CHOICE;
+  const isMulti = proposal?.account.voteType !== VoteType.SINGLE_CHOICE
+
+  const toggleShowNfts = () => {
+    setShowNfts(!showNfts)
+  }
+
+  useEffect(() => {
```
Comment (asktree) [inline pages/dao/[symbol]/proposal/[pk]/explore.tsx:46]: ? ? ?

const isNftMove = currentPluginPk &&
      NFT_PLUGINS_PKS.includes(currentPluginPk?.toBase58())

File: pages/dao/[symbol]/proposal/[pk]/explore.tsx [hunk 21]
```diff
@@ -53,19 +78,47 @@ export default function Explore() {
...
-          <h3 className="mb-4 mt-16">Top Voters</h3>
+          <div className="mb-4 mt-16 flex justify-between">
+            <h3 className="">Top Voters</h3>
+            <div
+              className={`${
+                isNftMode ? 'visible' : 'hidden'
+              } flex items-center`}
+            >
+              <p className="mb-0 mr-1 text-fgd-3">Show NFTs</p>
+              <Switch
```
Comment (asktree) [inline pages/dao/[symbol]/proposal/[pk]/explore.tsx:89]: meh just show nfts if its nft mode, why not?

File: pages/dao/[symbol]/proposal/[pk]/explore.tsx [hunk 22]
```diff
@@ -53,19 +78,47 @@ export default function Explore() {
...
+            <div className="flex flex-col gap-5 h-[500px]">
+              <ProposalTopVotersList
+                className={showNfts ? 'h-[275px]' : 'h-[500px]'}
+                data={records}
+                endpoint={endpoint}
+                isMulti={isMulti}
+                highlighted={highlighted}
+                onHighlight={setHighlighted}
+              />
+              <ProposalTopVotersNftChart
```
Comment (asktree) [inline pages/dao/[symbol]/proposal/[pk]/explore.tsx:110]: a comment here explaining the difference between this and `ProposalTopVotersList` maybe would be nice

File: utils/instructions/NftVoter/updateVoterWeight.ts [hunk 23]
```diff
@@ -0,0 +1,182 @@
...
+import { Program } from '@project-serum/anchor'
+
+type UpdateVoterWeightRecordTypes =
+  | 'castVote'
+  | 'commentProposal'
+  | 'createGovernance'
+  | 'createProposal'
+  | 'signOffProposal'
+
+class AccountData {
```
Comment (asktree) [inline utils/instructions/NftVoter/updateVoterWeight.ts:22]: ??

File: utils/instructions/NftVoter/castNftVote.ts [hunk 24]
```diff
@@ -0,0 +1,252 @@
...
+import { Program } from '@project-serum/anchor'
+
+type UpdateVoterWeightRecordTypes =
+  | 'castVote'
+  | 'commentProposal'
+  | 'createGovernance'
+  | 'createProposal'
+  | 'signOffProposal'
+
+class AccountData {
```
Comment (asktree) [inline utils/instructions/NftVoter/castNftVote.ts:26]: ??

File: actions/chat/postMessage.ts (hunk 5 above)
Comment (hauchu1998) [reply actions/chat/postMessage.ts:78]: Isn't the postMessageIxsChunk will happen after all the parallel txs are done?

File: hub/components/NewWallet/useNewWalletTransaction.ts (hunk 14 above)
Comment (hauchu1998) [reply hub/components/NewWallet/useNewWalletTransaction.ts:59]: hmm i thought the sequential one will wait for the parallel ones, not very sure as well. Since their document said we can create [parallel, sequential, parallel], and if the sequential won't wait, the original nft-voter castVote might not work well?

File: actions/chat/postMessage.ts (hunk 5 above)
Comment (asktree) [reply actions/chat/postMessage.ts:80]: the short answer is that i dont know and i find the mangolana source code too hard to follow. you should ask Adrian

File: hub/components/NewWallet/useNewWalletTransaction.ts (hunk 14 above)
Comment (asktree) [reply hub/components/NewWallet/useNewWalletTransaction.ts]: hmmmmm I don't know sorry.

File: utils/instructions/NftVoter/castNftVote.ts [hunk 25]
```diff
@@ -0,0 +1,234 @@
...
+import { NftVoterV2 } from 'idls/nft_voter_v2'
+import { Program } from '@project-serum/anchor'
+import {
+  AccountData,
+  UpdateVoterWeightRecordTypes,
+} from '@utils/uiTypes/VotePlugin'
+import { getCnftParamAndProof } from 'NftVotePlugin/getCnftParamAndProof'
+
+export const getCastNftVoteInstruction = async (
+  client: NftVoterClient,
```
Comment (asktree) [inline utils/instructions/NftVoter/castNftVote.ts:24]: we only ever use the `program` from the `client`, so that should be the parameter, not the client. I'm especially saying this because one day the vote client class / concept needs to be refactored.

now, the `program` object may not even be stateful, so it's unclear that even needs to be a parameter itself. but whatever, that could change in other versions of anchor maybe.

File: utils/instructions/NftVoter/updateVoterWeight.ts [hunk 26]
```diff
@@ -0,0 +1,164 @@
...
+import { NftVoterV2 } from 'idls/nft_voter_v2'
+import { Program } from '@project-serum/anchor'
+import {
+  AccountData,
+  UpdateVoterWeightRecordTypes,
+} from '@utils/uiTypes/VotePlugin'
+import { getCnftParamAndProof } from 'NftVotePlugin/getCnftParamAndProof'
+
+export const getUpdateVoterWeightRecordInstruction = async (
+  client: NftVoterClient,
```
Comment (asktree) [inline utils/instructions/NftVoter/updateVoterWeight.ts:20]: see other comment re: client

File: hub/components/NewWallet/useNewWalletTransaction.ts [hunk 27]
```diff
@@ -79,10 +84,25 @@ const useNewWalletCallback = (
       wallet.publicKey,
     );
 
+    // createTicketIxs is a list of instructions that create nftActionTicket only for nft-voter-v2 plugin
+    // so it will be empty for other plugins
+    const createTicketIxs = instructions.slice(0, -3);
```
Comment (asktree) [inline hub/components/NewWallet/useNewWalletTransaction.ts:89]: this kind of thing is a big no-no. makes the code really fragile, since adding instructions will break this. i understand why you did it but the old way was better than this. my comments before about changing the API were before I understood the parallel vs sequential thing. just do it how you were doing before. it's not your fault everything sucks. but yeah this is way too fragile.

**Comment by hauchu1998** [comment]
Test nft-voter-v1 (w/o supporting cNFT) in Preview
:white_check_mark:  Create a registrar
:white_check_mark:  Create a proposal
:white_check_mark:  Cast NFT votes
:white_check_mark:  Cast NFT vote with commenting
:white_check_mark:  Post messages
:white_check_mark:  Create new wallet(governance)
:white_check_mark:  Proposal stats page
:white_check_mark:  Member stats page
